lib/liquid/context.rb

```diff
@@ -207,9 +202,9 @@ def lookup_and_evaluate(obj, key, raise_on_not_found: true)
       value = obj[key]
 
       if value.is_a?(Proc) && obj.respond_to?(:[]=)
-        obj[key] = value.arity == 0 ? value.call : value.call(self)
+        obj[key] = contextualize(value)
```

Comment:
This would have been clearer, but achieve the same behaviour with more code. No strong opinion on my end.
```suggestion
        value = contextualize(value)
        obj[key] = value
        value
```